python/ray/serve/_private/application_state.py

```diff
@@ -27,61 +31,90 @@ def __init__(
             deployment_state_manager: deployment state manager which is used for
                 fetching deployment information
-            deploy_obj_ref: Task ObjRef of deploying application.
             deployment_time: Deployment timestamp
         """
-        if deploy_obj_ref:
-            self.status: ApplicationStatus = ApplicationStatus.DEPLOYING
-        else:
-            self.status: ApplicationStatus = ApplicationStatus.NOT_STARTED
+        self.status: ApplicationStatus = ApplicationStatus.NOT_STARTED
         self.name = name
         self.deployment_params: List[Dict] = []
         self.ready_to_be_deleted = False
         self.deployment_state_manager = deployment_state_manager
+        self.endpoint_state = endpoint_state
         if deployment_time:
             self.deployment_timestamp = deployment_time
         else:
             self.deployment_timestamp = time.time()
-        self.deploy_obj_ref = deploy_obj_ref
         self.app_msg = ""
         self.route_prefix = None
         self.docs_path = None
 
         # This set tracks old deployments that are being deleted
         self.deployments_to_delete = set()
 
+    def delete_deployment(self, name):
+        self.endpoint_state.delete_endpoint(name)
+        self.deployment_state_manager.delete_deployment(name)
+
     def delete(self):
         """Delete the application"""
         self.status = ApplicationStatus.DELETING
+        for deployment in self.deployments:
+            self.delete_deployment(deployment)
+
+    def set_deploy_failed(self, message):
+        self.status = ApplicationStatus.DEPLOY_FAILED
+        self.app_msg = message
+
+    def deploy_deployment(
+        self,
+        name: str,
+        deployment_config_proto_bytes: bytes,
+        replica_config_proto_bytes: bytes,
+        route_prefix: Optional[str],
+        deployer_job_id: Union[str, bytes],
+        docs_path: Optional[str] = None,
+        is_driver_deployment: Optional[bool] = False,
+    ) -> bool:
+        """
+        Deploys a deployment in the application.
 
-    def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        Returns: Whether the deployment is being updated.
+        """
+
+        updating = self.deployment_state_manager.deploy(
+            deployment_name=name,
+            deployment_config_proto_bytes=deployment_config_proto_bytes,
+            replica_config_proto_bytes=replica_config_proto_bytes,
+            deployer_job_id=deployer_job_id,
+            is_driver_deployment=is_driver_deployment,
+        )
+
+        if route_prefix is not None:
+            self.route_prefix = route_prefix
+            endpoint_info = EndpointInfo(route=route_prefix)
+            self.endpoint_state.update_endpoint(name, endpoint_info)
+        else:
+            self.endpoint_state.delete_endpoint(name)
+
+        return updating
+
+    def deploy(self, deployment_params: List[Dict]) -> List[bool]:
         """Deploy the application.
```

Comment:
please make this docstring more descriptive

also, nit: `apply` would be a more fitting name because this is deploying new deployments, updating existing ones, and deleting existing ones